swim/join_delayer.go

```diff
@@ -0,0 +1,204 @@
+func newDelayOpts() *delayOpts {
+	return &delayOpts{
+		initial:    defaultInitial,
+		max:        defaultMax,
+		randomizer: defaultRandomizer,
+		sleeper:    defaultSleeper,
+	}
+}
+
+// exponentialDelayer applies a delay in between repeated join attempts.
+// The delay increases exponentially and is capped at maxDelay.
+type exponentialDelayer struct {
+	// joiner is the address of the node sending join requests.
+	joiner string
+
+	// logger logs when a maximum delay is reached.
+	logger bark.Logger
+
+	// initialDelay is the fixed portion of the delay by which
+	// the exponential backoff is multiplied.
+	initialDelay float64
+
+	// maxDelay is the maximum delay applied to a join attempt.
+	maxDelay float64
+
+	// maxDelayReached is a flag that is toggled once the delay
+	// applied to a join attempt reaches or exceeds the max delay.
+	maxDelayReached bool
+
+	// nextDelayMin tracks the last used upper-bound for a join attempt delay.
+	// Upon the next join attempt, nextDelayMin will be used as the lower-bound
+	// for that delay. This acts as a shifting window for the bounds of the
+	// exponential backoff.
+	nextDelayMin float64
+
+	// randomizer generates a random delay in between a nextDelayMin
+	// and maxDelay.
+	randomizer delayRandomizer
+
+	// sleeper pauses execution of a join attempt.
+	sleeper delaySleeper
+}
+
+// newExponentialDelayer creates a new exponential delayer. joiner and logger
+// are required. opts is optional.
+func newExponentialDelayer(joiner string, logger bark.Logger,
+	opts *delayOpts) (*exponentialDelayer, error) {
+	if logger == nil {
+		return nil, errLoggerRequired
+	}
+
+	if opts == nil {
+		opts = newDelayOpts()
+	}
+
+	randomizer := opts.randomizer
+	if randomizer == nil {
+		randomizer = defaultRandomizer
+	}
+
+	sleeper := opts.sleeper
+	if sleeper == nil {
+		sleeper = defaultSleeper
+	}
+
+	return &exponentialDelayer{
+		joiner:          joiner,
+		logger:          logger,
+		initialDelay:    opts.initial,
+		nextDelayMin:    0,
+		maxDelayReached: false,
+		maxDelay:        opts.max,
+		randomizer:      randomizer,
+		sleeper:         sleeper,
+	}, nil
+}
+
+// nullDelayer is an empty implementation of joinDelayer.
+type nullDelayer struct {
+}
+
+// delay applies no delay.
+func (d *nullDelayer) delay(attemptNumber int) (time.Duration, error) {
+	return time.Duration(0), nil
+}
+
+// Delay delays a join attempt by sleeping for an amount of time. The
+// amount of time is computed as an exponential backoff based on the number
+// of join attempts that have been made at the time of the function call;
+// the number of attempts is 0-based. It returns a time.Duration equal to the
+// amount of delay applied and an error if one occurs. If an error occurs, a
+// duration equal to 0 will be returned, signifying no delay has been applied.
+func (d *exponentialDelayer) delay(attemptNumber int) (time.Duration, error) {
+	if attemptNumber < 0 {
```

Comment:
This is a good question. I like the idea of taking attemptNumber out of the signature.